python/ccf/ledger.py

```diff
@@ -303,13 +503,50 @@ def __init__(self, directory: str):
                     LOG.warning(f"The file {chunk} has not been committed")
                 self._filenames.append(os.path.join(directory, chunk))
 
+        # Initialize LedgerValidator isntance which will be passed to LedgerChunks.
+        self._ledger_validator = LedgerValidator()
+
     def __next__(self) -> LedgerChunk:
         self._fileindex += 1
         if len(self._filenames) > self._fileindex:
-            self._current_chunk = LedgerChunk(self._filenames[self._fileindex])
+            self._current_chunk = LedgerChunk(
+                self._filenames[self._fileindex], self._ledger_validator
+            )
             return self._current_chunk
         else:
+            LOG.info(
+                f"Ledger verification complete. Found {self._ledger_validator.signature_count} signatures."
+                + f"Ledger verified till seqNo: {self._ledger_validator.last_verified_seqno} and view: {self._ledger_validator.last_verified_view}"
+            )
             raise StopIteration
 
     def __iter__(self):
         return self
+
+
+def extract_msgpacked_data(data: bytes):
+    return msgpack.unpackb(data, **UNPACK_ARGS)
+
+
+class InvalidRootException(Exception):
+    """When the transactions don't add up to the ledger signature root"""
+
+
+class SignatureCountException(Exception):
+    """When a ledger has no signature transactions"""
+
+
+class SignatureEndingException(Exception):
+    """When a signature transaction is not the last transaction of a ledger"""
+
+
+class InvalidRootSignatureException(Exception):
+    """When a signature transactions certificate cannot be validated against the node signature"""
```

Comment:
Can we change the tense of these to match the `Exception`s below (tell the user what went wrong, rather than describing when they would be thrown?

```suggestion
class InvalidRootException(Exception):
    """Merkle root constructed from transaction digests doesn't match the root in signature transaction"""


class SignatureCountException(Exception):
    """The ledger contains no signature transactions"""


class SignatureEndingException(Exception):
    """The last transaction in the ledger is not a signature transaction"""


class InvalidRootSignatureException(Exception):
    """Signature transaction contains invalid signature"""
```